linalg/auxiliary.cpp

```diff
@@ -0,0 +1,542 @@
+MatrixFreeAuxiliarySpace::MatrixFreeAuxiliarySpace(
+   ParMesh& mesh_lor, Coefficient* alpha_coeff,
+   Coefficient* beta_coeff, MatrixCoefficient* beta_mcoeff, Array<int>& ess_bdr,
+   Operator& curlcurl_oper, Operator& pi,
+#ifdef MFEM_USE_AMGX
+   bool useAmgX_,
+#endif
+   int cg_iterations) :
+   Solver(pi.Width()),
+   comm(mesh_lor.GetComm()),
+   matfree(NULL),
+   cg(NULL),
+#ifdef MFEM_USE_AMGX
+   useAmgX(useAmgX_),
+#endif
+   inner_aux_iterations(0)
+{
+   H1_FECollection * fec_lor = new H1_FECollection(1, mesh_lor.Dimension());
+   ParFiniteElementSpace fespace_lor_d(&mesh_lor, fec_lor, mesh_lor.Dimension(),
+                                       Ordering::byVDIM);
+
+   // build LOR AMG v-cycle
+   if (ess_bdr.Size())
+   {
+      fespace_lor_d.GetEssentialTrueDofs(ess_bdr, ess_tdof_list);
+   }
+   ParBilinearForm a_space(&fespace_lor_d);
+
+   // this choice of policy is super-important for the G-space solver, but
+   // also can make some difference here
+   const Matrix::DiagonalPolicy policy = Matrix::DIAG_KEEP;
+   a_space.SetDiagonalPolicy(policy); // doesn't do anything, see Eliminate() below
+   if (alpha_coeff == NULL)
+   {
+      a_space.AddDomainIntegrator(new VectorDiffusionIntegrator);
+   }
+   else
+   {
+      a_space.AddDomainIntegrator(new VectorDiffusionIntegrator(*alpha_coeff));
+   }
+
+   if (beta_mcoeff != NULL)
+   {
+      MFEM_VERIFY(beta_coeff == NULL, "Only one beta coefficient should be defined.");
+      a_space.AddDomainIntegrator(new VectorMassIntegrator(*beta_mcoeff));
+   }
+   else if (beta_coeff != NULL)
+   {
+      a_space.AddDomainIntegrator(new VectorMassIntegrator(*beta_coeff));
+   }
+   else
+   {
+      a_space.AddDomainIntegrator(new VectorMassIntegrator);
+   }
+   a_space.UsePrecomputedSparsity();
+   a_space.Assemble();
+   a_space.EliminateEssentialBC(ess_bdr, policy);
+   a_space.Finalize();
+   aspacematrix = a_space.ParallelAssemble();
+   aspacematrix->CopyRowStarts();
+   aspacematrix->CopyColStarts();
+
+   SetupAMG(fespace_lor_d.GetMesh()->Dimension());
+
+   if (cg_iterations > 0)
+   {
+      SetupCG(curlcurl_oper, pi, cg_iterations);
+   }
+   else
+   {
+      SetupVCycle();
+   }
+   delete fec_lor;
+}
+
+/* G-space constructor
+
+   The auxiliary space solves in general, and this one in particular,
+   seem to be quite sensitive to handling of boundary conditions. Note
+   some careful choices for Matrix::DiagonalPolicy and the ZeroWrap
+   object, as well as the use of a single CG iteration (instead of just
+   an AMG V-cycle). */
+MatrixFreeAuxiliarySpace::MatrixFreeAuxiliarySpace(
+   ParMesh& mesh_lor, Coefficient* beta_coeff,
+   MatrixCoefficient* beta_mcoeff, Array<int>& ess_bdr, Operator& curlcurl_oper,
+   Operator& g,
+#ifdef MFEM_USE_AMGX
+   bool useAmgX_,
+#endif
+   int cg_iterations)
+   :
+   Solver(curlcurl_oper.Height()),
+   comm(mesh_lor.GetComm()),
+   matfree(NULL),
+   cg(NULL),
+#ifdef MFEM_USE_AMGX
+   useAmgX_(useAmgX),
```

Comment:
It looks like the underscores got mixed up.

Reply:
Good catch, fixed.